// @forestadmin/datasource-customizer/dist/decorators/validation/collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
class ValidationDecorator extends datasource_toolkit_1.CollectionDecorator {
    constructor() {
        super(...arguments);
        this.validation = {};
    }
    addValidation(name, validation) {
        var _a;
        datasource_toolkit_1.FieldValidator.validate(this, name);
        const field = this.childCollection.schema.fields[name];
        if (field?.type !== 'Column')
            throw new Error('Cannot add validators on a relation, use the foreign key instead');
        if (field.isReadOnly)
            throw new Error('Cannot add validators on a readonly field');
        (_a = this.validation)[name] ?? (_a[name] = []);
        this.validation[name].push(validation);
        this.markSchemaAsDirty();
    }
    async create(caller, data) {
        for (const record of data)
            this.validate(record, caller.timezone, true);
        return super.create(caller, data);
    }
    async update(caller, filter, patch) {
        this.validate(patch, caller.timezone, false);
        return super.update(caller, filter, patch);
    }
    refineSchema(subSchema) {
        const schema = { ...subSchema, fields: { ...subSchema.fields } };
        for (const [name, rules] of Object.entries(this.validation)) {
            const field = { ...schema.fields[name] };
            field.validation = [...(field.validation ?? []), ...rules];
            schema.fields[name] = field;
        }
        return schema;
    }
    validate(record, timezone, allFields) {
        for (const [name, rules] of Object.entries(this.validation)) {
            if (allFields || record[name] !== undefined) {
                // When setting a field to null, only the "Present" validator is relevant
                const applicableRules = record[name] === null ? rules.filter(r => r.operator === 'Present') : rules;
                for (const validator of applicableRules) {
                    const rawLeaf = { field: name, ...validator };
                    const tree = datasource_toolkit_1.ConditionTreeFactory.fromPlainObject(rawLeaf);
                    if (!tree.match(record, this, timezone)) {
                        const message = `'${name}' failed validation rule:`;
                        const rule = validator.value !== undefined
                            ? `${validator.operator}(${validator.value})`
                            : `${validator.operator}`;
                        throw new datasource_toolkit_1.ValidationError(`${message} '${rule}'`);
                    }
                }
            }
        }
    }
}
exports.default = ValidationDecorator;
