// @forestadmin/forestadmin-client/dist/events-subscription/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const eventsource_1 = __importDefault(require("eventsource"));
const types_1 = require("./types");
class EventsSubscriptionService {
    constructor(options, refreshEventsHandlerService) {
        this.options = options;
        this.refreshEventsHandlerService = refreshEventsHandlerService;
    }
    async subscribeEvents() {
        if (!this.options.instantCacheRefresh) {
            this.options.logger('Debug', 'Event source deactivated.. Use agent option [instantCacheRefresh=true] ' +
                'if you want to activate them');
            return;
        }
        const eventSourceConfig = {
            // forest-secret-key act as the credential
            withCredentials: false,
            headers: { 'forest-secret-key': this.options.envSecret },
            https: {
                rejectUnauthorized: process.env.NODE_TLS_REJECT_UNAUTHORIZED !== '0',
            },
        };
        const url = new URL('/liana/v4/subscribe-to-events', this.options.forestServerUrl).toString();
        const eventSource = new eventsource_1.default(url, eventSourceConfig);
        // Override reconnect interval to 5 seconds
        eventSource.reconnectInterval = 5000;
        eventSource.addEventListener('error', this.onEventError.bind(this));
        // Only listen after first open
        eventSource.once('open', () => eventSource.addEventListener('open', () => this.onEventOpenAgain()));
        eventSource.addEventListener(types_1.ServerEventType.RefreshUsers, async () => this.refreshEventsHandlerService.refreshUsers());
        eventSource.addEventListener(types_1.ServerEventType.RefreshRoles, async () => this.refreshEventsHandlerService.refreshRoles());
        eventSource.addEventListener(types_1.ServerEventType.RefreshRenderings, async (event) => this.handleSeverEventRefreshRenderings(event));
        eventSource.addEventListener(types_1.ServerEventType.RefreshCustomizations, async () => this.refreshEventsHandlerService.refreshCustomizations());
        this.eventSource = eventSource;
    }
    /**
     * Close the current EventSource
     */
    close() {
        this.eventSource?.close();
    }
    async handleSeverEventRefreshRenderings(event) {
        if (!event.data) {
            this.options.logger('Debug', 'Server Event - RefreshRenderings missing required data.');
            return;
        }
        const { renderingIds } = JSON.parse(event.data);
        await this.refreshEventsHandlerService.refreshRenderings(renderingIds);
    }
    onEventError(event) {
        const { status, message } = event;
        if ([502, 503, 504].includes(status)) {
            this.options.logger('Debug', 'Server Event - Connection lost trying to reconnect…');
            return;
        }
        if (status === 404)
            throw new Error('Forest Admin server failed to find the environment ' +
                'related to the envSecret you configured. ' +
                'Can you check that you copied it properly during initialization?');
        if (message)
            this.options.logger('Warn', `Server Event - Error: ${JSON.stringify(event)}`);
    }
    onEventOpenAgain() {
        this.options.logger('Debug', 'Server Event - Open EventSource (SSE) connection with Forest Admin servers');
        // Flush all previous data as we could have missed some events
        this.refreshEventsHandlerService.refreshEverything();
    }
}
exports.default = EventsSubscriptionService;
