// @forestadmin/forestadmin-client/dist/events-subscription/native-refresh-events-handler-service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = __importDefault(require("events"));
class NativeRefreshEventsHandlerService {
    constructor(actionPermissionService, usersPermissionService, renderingPermissionService) {
        this.actionPermissionService = actionPermissionService;
        this.usersPermissionService = usersPermissionService;
        this.renderingPermissionService = renderingPermissionService;
        this.eventEmitter = new events_1.default();
    }
    refreshUsers() {
        this.usersPermissionService.invalidateCache();
    }
    refreshRoles() {
        this.actionPermissionService.invalidateCache();
    }
    refreshRenderings(renderingIds) {
        for (const renderingId of renderingIds)
            this.renderingPermissionService.invalidateCache(renderingId);
    }
    refreshCustomizations() {
        this.eventEmitter.emit('RefreshCustomizations');
    }
    onRefreshCustomizations(handler) {
        this.eventEmitter.on('RefreshCustomizations', handler);
    }
    refreshEverything() {
        this.usersPermissionService.invalidateCache();
        this.actionPermissionService.invalidateCache();
        this.renderingPermissionService.invalidateAllCache();
        // Emit RefreshCustomizations event
        this.eventEmitter.emit('RefreshCustomizations');
    }
}
exports.default = NativeRefreshEventsHandlerService;
