// @forestadmin/forestadmin-client/dist/auth/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const openid_client_1 = require("openid-client");
const server_1 = __importDefault(require("../utils/server"));
class AuthService {
    constructor(options) {
        this.options = options;
    }
    async getOpenIdClient() {
        // We can't use 'Issuer.discover' because the oidc config is behind an auth-wall.
        const url = '/oidc/.well-known/openid-configuration';
        const config = await server_1.default.query(this.options, 'get', url);
        const issuer = new openid_client_1.Issuer(config);
        const registration = { token_endpoint_auth_method: 'none' };
        return issuer.Client.register(registration, {
            initialAccessToken: this.options.envSecret,
        });
    }
    async getUserInfo(renderingId, accessToken) {
        const url = `/liana/v2/renderings/${renderingId}/authorization`;
        const headers = { 'forest-token': accessToken };
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const response = await server_1.default.query(this.options, 'get', url, headers);
        return {
            id: Number(response.data.id),
            email: response.data.attributes.email,
            firstName: response.data.attributes.first_name,
            lastName: response.data.attributes.last_name,
            team: response.data.attributes.teams[0],
            role: response.data.attributes.role,
            permissionLevel: response.data.attributes.permission_level,
            renderingId,
            tags: response.data.attributes.tags?.reduce((memo, { key, value }) => ({ ...memo, [key]: value }), {}),
        };
    }
}
exports.default = AuthService;
