// @forestadmin/datasource-customizer/dist/decorators/publication/collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
/** This decorator allows hiding fields */
class PublicationCollectionDecorator extends datasource_toolkit_1.CollectionDecorator {
    constructor() {
        super(...arguments);
        this.blacklist = new Set();
    }
    /** Show/hide fields from the schema */
    changeFieldVisibility(name, visible) {
        const field = this.childCollection.schema.fields[name];
        if (!field) {
            throw new Error(`No such field '${name}'`);
        }
        if (datasource_toolkit_1.SchemaUtils.isPrimaryKey(this.childCollection.schema, name)) {
            throw new Error(`Cannot hide primary key`);
        }
        if (!visible)
            this.blacklist.add(name);
        else
            this.blacklist.delete(name);
        this.markSchemaAsDirty();
    }
    async create(caller, data) {
        const records = await super.create(caller, data);
        return records.map(childRecord => {
            const record = {};
            for (const key of Object.keys(childRecord))
                if (!this.blacklist.has(key))
                    record[key] = childRecord[key];
            return record;
        });
    }
    refineSchema(childSchema) {
        const fields = {};
        for (const [name, field] of Object.entries(childSchema.fields)) {
            if (this.isPublished(name)) {
                fields[name] = field;
            }
        }
        return { ...childSchema, fields };
    }
    isPublished(name) {
        // Explicitly hidden
        if (this.blacklist.has(name))
            return false;
        // Implicitly hidden
        const field = this.childCollection.schema.fields[name];
        if (field.type === 'ManyToOne')
            return (this.dataSource.isPublished(field.foreignCollection) &&
                this.isPublished(field.foreignKey) &&
                this.dataSource.getCollection(field.foreignCollection).isPublished(field.foreignKeyTarget));
        if (field.type === 'OneToOne' || field.type === 'OneToMany')
            return (this.dataSource.isPublished(field.foreignCollection) &&
                this.dataSource.getCollection(field.foreignCollection).isPublished(field.originKey) &&
                this.isPublished(field.originKeyTarget));
        if (field.type === 'ManyToMany')
            return (this.dataSource.isPublished(field.throughCollection) &&
                this.dataSource.isPublished(field.foreignCollection) &&
                this.dataSource.getCollection(field.throughCollection).isPublished(field.foreignKey) &&
                this.dataSource.getCollection(field.throughCollection).isPublished(field.originKey) &&
                this.isPublished(field.originKeyTarget) &&
                this.dataSource.getCollection(field.foreignCollection).isPublished(field.foreignKeyTarget));
        return true;
    }
    markSchemaAsDirty() {
        return super.markSchemaAsDirty();
    }
}
exports.default = PublicationCollectionDecorator;
