// @forestadmin/datasource-customizer/dist/decorators/publication/datasource.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const collection_1 = __importDefault(require("./collection"));
class PublicationDataSourceDecorator extends datasource_toolkit_1.DataSourceDecorator {
    constructor(childDataSource) {
        super(childDataSource, collection_1.default);
        this.blacklist = new Set();
    }
    get collections() {
        return this.childDataSource.collections
            .filter(({ name }) => !this.blacklist.has(name))
            .map(({ name }) => this.getCollection(name));
    }
    getCollection(name) {
        if (this.blacklist.has(name))
            throw new Error(`Collection "${name}" was removed.`);
        return super.getCollection(name);
    }
    keepCollectionsMatching(include, exclude) {
        this.validateCollectionNames([...(include ?? []), ...(exclude ?? [])]);
        // List collection we're keeping from the white/black list.
        for (const { name } of this.collections) {
            if ((include && !include.includes(name)) || exclude?.includes(name)) {
                this.removeCollection(name);
            }
        }
    }
    removeCollection(collectionName) {
        this.validateCollectionNames([collectionName]);
        // Delete the collection
        this.blacklist.add(collectionName);
        // Tell all collections that their schema is dirty: if we removed a collection, all
        // relations to this collection are now invalid and should be unpublished.
        for (const collection of this.collections) {
            collection.markSchemaAsDirty();
        }
    }
    validateCollectionNames(names) {
        for (const name of names)
            this.getCollection(name);
    }
    isPublished(collectionName) {
        return !this.blacklist.has(collectionName);
    }
}
exports.default = PublicationDataSourceDecorator;
