// @forestadmin/datasource-customizer/dist/decorators/rename-collection/datasource.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const collection_1 = __importDefault(require("./collection"));
class RenameCollectionDataSourceDecorator extends datasource_toolkit_1.DataSourceDecorator {
    constructor(childDataSource) {
        super(childDataSource, collection_1.default);
        this.fromChildName = {};
        this.toChildName = {};
    }
    get collections() {
        return this.childDataSource.collections.map(({ name }) => super.getCollection(name));
    }
    getCollection(name) {
        // Collection has been renamed, user is using the new name
        if (this.toChildName[name]) {
            return super.getCollection(this.toChildName[name]);
        }
        // Collection has been renamed, user is using the old name
        if (this.fromChildName[name]) {
            throw new Error(`Collection '${name}' has been renamed to '${this.fromChildName[name]}'`);
        }
        // Collection has not been renamed
        return super.getCollection(name);
    }
    /**
     * Helper to rename multiple collections at once.
     * Used by the addDataSource method in DataSourceCustomizer.
     */
    renameCollections(rename) {
        const tuples = typeof rename === 'function'
            ? this.collections.map(({ name }) => [name, rename(name) ?? name])
            : Object.entries(rename ?? {});
        for (const [oldName, newName] of tuples) {
            this.renameCollection(oldName, newName);
        }
    }
    /**
     * Rename a single collection
     * Used by the renameCollection method in DataSourceCustomizer.
     * @param currentName
     * @param newName
     */
    renameCollection(currentName, newName) {
        // Check collection exists
        this.getCollection(currentName);
        // Rename collection
        if (currentName !== newName) {
            // Check new name is not already used
            if (this.collections.some(({ name }) => name === newName)) {
                throw new Error(`The given new collection name "${newName}" is already defined`);
            }
            // Check we don't rename a collection twice
            if (this.toChildName[currentName]) {
                throw new Error(`Cannot rename a collection twice: ` +
                    `${this.toChildName[currentName]}->${currentName}->${newName}`);
            }
            this.fromChildName[currentName] = newName;
            this.toChildName[newName] = currentName;
            for (const collection of this.collections) {
                collection.markSchemaAsDirty();
            }
        }
    }
    /** @internal */
    getCollectionName(childName) {
        return this.fromChildName[childName] ?? childName;
    }
}
exports.default = RenameCollectionDataSourceDecorator;
