// @forestadmin/forestadmin-client/dist/model-customizations/model-customization-from-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("./types");
function mapApiValues(modelCustomization) {
    switch (modelCustomization.type) {
        case types_1.ModelCustomizationType.action: {
            const configuration = modelCustomization.configuration;
            const mappedConfiguration = {
                ...configuration,
                scope: configuration.scope
                    ? (configuration.scope.slice(0, 1).toUpperCase() +
                        configuration.scope.slice(1))
                    : 'Single',
            };
            return {
                ...modelCustomization,
                configuration: mappedConfiguration,
            };
        }
        default:
            throw new Error('Only action customizations are supported for now.');
    }
}
class ModelCustomizationFromApiService {
    constructor(forestadminServerInterface, options) {
        this.forestadminServerInterface = forestadminServerInterface;
        this.options = options;
    }
    async getConfiguration() {
        const result = await this.forestadminServerInterface.getModelCustomizations(this.options);
        return result.map(mapApiValues);
    }
}
exports.default = ModelCustomizationFromApiService;
