// @forestadmin/datasource-toolkit/dist/src/interfaces/query/condition-tree/nodes/leaf.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_1 = __importDefault(require("./base"));
const operators_1 = require("./operators");
const collection_1 = __importDefault(require("../../../../utils/collection"));
const record_1 = __importDefault(require("../../../../utils/record"));
const projection_1 = __importDefault(require("../../projection"));
const equivalence_1 = __importDefault(require("../equivalence"));
const factory_1 = __importDefault(require("../factory"));
class ConditionTreeLeaf extends base_1.default {
    get projection() {
        return new projection_1.default(this.field);
    }
    get useIntervalOperator() {
        return operators_1.intervalOperators.includes(this.operator);
    }
    constructor(field, operator, value) {
        super();
        this.field = field;
        this.operator = operator;
        this.value = value;
    }
    forEachLeaf(handler) {
        handler(this);
    }
    everyLeaf(handler) {
        return handler(this);
    }
    someLeaf(handler) {
        return handler(this);
    }
    inverse() {
        if (operators_1.allOperators.includes(`Not${this.operator}`)) {
            return this.override({ operator: `Not${this.operator}` });
        }
        if (this.operator.startsWith('Not')) {
            return this.override({ operator: this.operator.substring(3) });
        }
        switch (this.operator) {
            case 'Blank':
                return this.override({ operator: 'Present' });
            case 'Present':
                return this.override({ operator: 'Blank' });
            default:
                throw new Error(`Operator '${this.operator}' cannot be inverted.`);
        }
    }
    replaceLeafs(handler, bind) {
        const result = handler.call(bind, this);
        return result instanceof base_1.default ? result : factory_1.default.fromPlainObject(result);
    }
    async replaceLeafsAsync(handler, bind) {
        const result = await handler.call(bind, this);
        return result instanceof base_1.default ? result : factory_1.default.fromPlainObject(result);
    }
    match(record, collection, timezone) {
        const fieldValue = record_1.default.getFieldValue(record, this.field);
        const { columnType } = collection_1.default.getFieldSchema(collection, this.field);
        const supported = [
            ...['In', 'Equal', 'LessThan', 'GreaterThan', 'Match', 'StartsWith', 'EndsWith'],
            ...['LongerThan', 'ShorterThan', 'IncludesAll', 'NotIn', 'NotEqual', 'NotContains'],
        ];
        switch (this.operator) {
            case 'In':
                return this.value?.includes(fieldValue);
            case 'Equal':
                return fieldValue == this.value; // eslint-disable-line eqeqeq
            case 'LessThan':
                return fieldValue < this.value;
            case 'GreaterThan':
                return fieldValue > this.value;
            case 'Match':
                return typeof fieldValue === 'string' && this.value.test(fieldValue);
            case 'StartsWith':
                return typeof fieldValue === 'string' && fieldValue.startsWith(this.value);
            case 'EndsWith':
                return typeof fieldValue === 'string' && fieldValue.endsWith(this.value);
            case 'LongerThan':
                return typeof fieldValue === 'string' ? fieldValue.length > this.value : false;
            case 'ShorterThan':
                return typeof fieldValue === 'string' ? fieldValue.length < this.value : false;
            case 'IncludesAll':
                return !!this.value?.every(v => fieldValue?.includes(v));
            case 'NotIn':
            case 'NotEqual':
            case 'NotContains':
                return !this.inverse().match(record, collection, timezone);
            default:
                return equivalence_1.default.getEquivalentTree(this, new Set(supported), columnType, timezone).match(record, collection, timezone);
        }
    }
    override(params) {
        return factory_1.default.fromPlainObject({ ...this, ...params });
    }
    unnest() {
        return super.unnest();
    }
}
exports.default = ConditionTreeLeaf;
