// @forestadmin/datasource-toolkit/dist/src/interfaces/query/condition-tree/transforms/comparison.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const factory_1 = __importDefault(require("../factory"));
const leaf_1 = __importDefault(require("../nodes/leaf"));
exports.default = () => ({
    Blank: [
        {
            dependsOn: ['In'],
            forTypes: ['String'],
            replacer: leaf => leaf.override({ operator: 'In', value: [null, ''] }),
        },
        {
            dependsOn: ['Missing'],
            replacer: leaf => leaf.override({ operator: 'Missing' }),
        },
    ],
    Missing: [
        {
            dependsOn: ['Equal'],
            replacer: leaf => leaf.override({ operator: 'Equal', value: null }),
        },
    ],
    Present: [
        {
            dependsOn: ['NotIn'],
            forTypes: ['String'],
            replacer: leaf => leaf.override({ operator: 'NotIn', value: [null, ''] }),
        },
        {
            dependsOn: ['NotEqual'],
            replacer: leaf => leaf.override({ operator: 'NotEqual', value: null }),
        },
    ],
    Equal: [
        {
            dependsOn: ['In'],
            replacer: leaf => leaf.override({ operator: 'In', value: [leaf.value] }),
        },
    ],
    In: [
        {
            dependsOn: ['Equal', 'Match'],
            forTypes: ['String'],
            replacer: leaf => {
                const values = leaf.value;
                const conditions = [];
                for (const value of [null, ''])
                    if (values.includes(value))
                        conditions.push(new leaf_1.default(leaf.field, 'Equal', value));
                if (values.some(v => v !== null && v !== '')) {
                    const escaped = values.filter(Boolean).map(str => str.replace(/[.|[\]]/g, m => `\\${m}`));
                    const regexp = new RegExp(`^${escaped.join('|')}$`, 'g');
                    conditions.push(new leaf_1.default(leaf.field, 'Match', regexp));
                }
                return factory_1.default.union(...conditions);
            },
        },
        {
            dependsOn: ['Equal'],
            replacer: leaf => factory_1.default.union(...leaf.value.map(item => leaf.override({ operator: 'Equal', value: item }))),
        },
    ],
    NotEqual: [
        {
            dependsOn: ['NotIn'],
            replacer: leaf => leaf.override({ operator: 'NotIn', value: [leaf.value] }),
        },
    ],
    NotIn: [
        {
            dependsOn: ['NotEqual', 'Match'],
            forTypes: ['String'],
            replacer: leaf => {
                const values = leaf.value;
                const conditions = [];
                for (const value of [null, ''])
                    if (values.includes(value))
                        conditions.push(new leaf_1.default(leaf.field, 'NotEqual', value));
                if (values.some(v => v !== null && v !== '')) {
                    const escaped = values.filter(Boolean).map(str => str.replace(/[.|[\]]/g, m => `\\${m}`));
                    const regexp = new RegExp(`(?!${escaped.join('|')})`, 'g');
                    conditions.push(new leaf_1.default(leaf.field, 'Match', regexp));
                }
                return factory_1.default.intersect(...conditions);
            },
        },
        {
            dependsOn: ['NotEqual'],
            replacer: leaf => factory_1.default.intersect(...leaf.value.map(item => leaf.override({ operator: 'NotEqual', value: item }))),
        },
    ],
});
