// @forestadmin/datasource-toolkit/dist/src/interfaces/query/condition-tree/transforms/pattern.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function likes(getPattern, caseSensitive) {
    const operator = caseSensitive ? 'Like' : 'ILike';
    return {
        dependsOn: [operator],
        forTypes: ['String'],
        replacer: leaf => leaf.override({ operator, value: getPattern(leaf.value) }),
    };
}
function match(caseSensitive) {
    return {
        dependsOn: ['Match'],
        forTypes: ['String'],
        replacer: leaf => {
            let regexp = leaf.value;
            // eslint-disable-next-line no-useless-escape
            regexp = regexp.replace(/([\.\\\+\*\?\[\^\]\$\(\)\{\}\=\!\<\>\|\:\-])/g, '\\$1');
            regexp = regexp.replace(/%/g, '.*').replace(/_/g, '.');
            return leaf.override({
                operator: 'Match',
                value: RegExp(`^${regexp}$`, caseSensitive ? 'g' : 'gi'),
            });
        },
    };
}
exports.default = () => ({
    Contains: [likes(value => `%${value}%`, true)],
    StartsWith: [likes(value => `${value}%`, true)],
    EndsWith: [likes(value => `%${value}`, true)],
    IContains: [likes(value => `%${value}%`, false)],
    IStartsWith: [likes(value => `${value}%`, false)],
    IEndsWith: [likes(value => `%${value}`, false)],
    ILike: [match(false)],
    Like: [match(true)],
});
