// @forestadmin/datasource-customizer/dist/decorators/actions/context/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable max-classes-per-file */
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const collection_context_1 = __importDefault(require("../../../context/collection-context"));
class ActionContext extends collection_context_1.default {
    /**
     * @deprecated use `hasFieldChange` instead. [linked issue](https://github.com/ForestAdmin/agent-nodejs/issues/815).
     * @todo remove accessor
     */
    get changedField() {
        console.warn('\x1b[33mwarning:\x1b[0m', 'Usage of `changedField` is deprecated, please use `hasFieldChanged` instead.');
        return this._changedField;
    }
    constructor(collection, caller, formValue, filter, used, changedField) {
        super(collection, caller);
        this.formValues = formValue;
        this.filter = filter;
        this._changedField = changedField;
        this.reset();
        // Spy on which formValues are accessed to set-up change hooks
        if (used) {
            this.formValues = new Proxy(this.formValues, {
                get: (target, prop, receiver) => {
                    if (typeof prop === 'string')
                        used.add(prop);
                    return Reflect.get(target, prop, receiver);
                },
                set: () => {
                    throw new Error('formValues is readonly');
                },
            });
            this.hasFieldChanged = (fieldName) => {
                used.add(fieldName);
                return this._changedField === fieldName;
            };
        }
    }
    /**
     * Get all the records selected by an action
     * @param fields An array of fields needed in the response
     * @example
     * .getRecords(['id', 'isActive', 'name']);
     */
    async getRecords(fields) {
        // This function just queues the request into this.queries, so that we can merge all calls
        // to getRecords() into a single one.
        // The call to setTimeout which resolve the promises will trigger only once all handlers in
        // the customer's form have been called as Promises are queued before calls to setTimeout
        // in Node.js event loop
        // @see https://dev.to/khaosdoctor/node-js-under-the-hood-3-deep-dive-into-the-event-loop-135d\
        //   #microtasks-and-macrotasks
        //   Ordering of micro/macro tasks in Node.js event loop
        //
        // @see https://github.com/graphql/dataloader
        //   A library from facebook from which this pattern is inspired.
        datasource_toolkit_1.ProjectionValidator.validate(this.realCollection, fields);
        const deferred = new datasource_toolkit_1.Deferred();
        const projection = new datasource_toolkit_1.Projection(...fields);
        if (this.queries.length === 0)
            setTimeout(() => this.runQuery());
        this.queries.push({ projection, deferred });
        this.projection = this.projection.union(projection);
        return deferred.promise;
    }
    /**
     * Get all the records ids selected by an action
     */
    async getRecordIds() {
        const compositeIds = await this.getCompositeRecordIds();
        return compositeIds.map(id => id[0]);
    }
    /**
     * Get all the records ids (when the collection uses composite keys)
     */
    async getCompositeRecordIds() {
        const projection = new datasource_toolkit_1.Projection().withPks(this.realCollection);
        const records = await this.getRecords(projection);
        return records.map(r => datasource_toolkit_1.RecordUtils.getPrimaryKey(this.realCollection.schema, r));
    }
    async runQuery() {
        const { queries, projection } = this;
        this.reset();
        try {
            // Run a single query which contains all fields / relations which were requested by
            // the different calls made to getRecords
            const records = await this.collection.list(this.filter, projection);
            // Resolve each on of the promises only with the requested fields.
            for (const query of queries)
                query.deferred.resolve(query.projection.apply(records));
        }
        catch (e) {
            // Rejecting each promises at next tick
            // This ensures that we don't let any promise hanging forever if the customer throws in
            // the rejection handler.
            for (const query of queries) {
                process.nextTick(() => query.deferred.reject(e));
            }
        }
    }
    reset() {
        this.queries = [];
        this.projection = new datasource_toolkit_1.Projection();
    }
}
exports.default = ActionContext;
