// @forestadmin/datasource-customizer/dist/decorators/composite-datasource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class CompositeDatasource {
    constructor() {
        this.dataSources = [];
    }
    get schema() {
        return { charts: this.dataSources.flatMap(dataSource => dataSource.schema.charts) };
    }
    get collections() {
        return this.dataSources.flatMap(dataSource => dataSource.collections);
    }
    getCollection(name) {
        for (const collection of this.dataSources) {
            try {
                return collection.getCollection(name);
            }
            catch (error) {
                // Ignore
            }
        }
        throw new Error(`Collection '${name}' not found. List of available collections: ${this.collections
            .map(c => c.name)
            .sort()
            .join(', ')}`);
    }
    addDataSource(dataSource) {
        for (const collection of dataSource.collections) {
            if (this.collections.some(c => c.name === collection.name)) {
                throw new Error(`Collection '${collection.name}' already exists`);
            }
        }
        for (const chart of dataSource.schema.charts) {
            if (this.schema.charts.includes(chart)) {
                throw new Error(`Chart '${chart}' is already defined in datasource.`);
            }
        }
        this.dataSources.push(dataSource);
    }
    async renderChart(caller, name) {
        for (const dataSource of this.dataSources) {
            if (dataSource.schema.charts.includes(name)) {
                return dataSource.renderChart(caller, name);
            }
        }
        throw new Error(`Chart '${name}' is not defined in the dataSource.`);
    }
}
exports.default = CompositeDatasource;
